resolver/caching_resolver.go

```diff
@@ -107,28 +108,36 @@ func configureCaches(ctx context.Context, c *CachingResolver, cfg *config.Cachin
 func (r *CachingResolver) reloadCacheEntry(ctx context.Context, cacheKey string) (*[]byte, time.Duration) {
 	qType, domainName := util.ExtractCacheKey(cacheKey)
 	ctx, logger := r.log(ctx)
+	logger = logger.WithField("domain", util.Obfuscate(domainName))
 
-	logger.Debugf("prefetching '%s' (%s)", util.Obfuscate(domainName), qType)
+	logger.Debugf("prefetching %s", qType)
 
 	req := newRequest(dns.Fqdn(domainName), qType)
+
 	response, err := r.next.Resolve(ctx, req)
+	if err != nil {
+		logger.WithError(err).Warn("cache prefetch failed")
 
-	if err == nil {
-		if response.Res.Rcode == dns.RcodeSuccess {
-			packed, err := response.Res.Pack()
-			if err != nil {
-				logger.Error("unable to pack response", err)
+		return nil, 0
+	}
 
-				return nil, 0
-			}
+	cacheCopy, ttl := r.createCacheEntry(logger, response.Res)
+	if cacheCopy == nil || ttl == noCacheTTL {
+		return nil, 0
+	}
 
-			return &packed, r.adjustTTLs(response.Res.Answer)
-		}
-	} else {
-		util.LogOnError(ctx, fmt.Sprintf("can't prefetch '%s' ", domainName), err)
+	packed, err := cacheCopy.Pack()
+	if err != nil {
+		logger.WithError(err).WithError(err).Warn("response packing failed")
+
+		return nil, 0
 	}
 
-	return nil, 0
+	if r.redisClient != nil {
+		r.redisClient.PublishCache(cacheKey, cacheCopy)
+	}
+
+	return &packed, util.ToTTLDuration(ttl)
```

Comment:
I think the biggest cause of dupe code is that here we need to return the new cache entry,
but in `r.Resolver` we need to call `r.resultCache.Put`.
So we could change the prefetch API to be more like the normal caching flow:
`PrefetchingOptions.ReloadFn func(ctx context.Context, key string) (*T, time.Duration)`
could be something like `OnAboutToExpire func(ctx context.Context, key string)`,
where the callback is expected to call `Put` if they want to update the value.
That way `reloadCacheEntry` could become a tiny wrapper around `r.Resolve` (or a new `resolve` method they both call).

From a quick check, `ReloadFn` is only used by this code (and the cache tests),
and `Put` allows replacing a value, so I think that should be doable!

From there I think most things will fall into place :)

Reply:
Thanks for the input I think that was the part which caused me the most headaches. 😅

I also try to refactor the caches to streamline the resolver behavior. 👍